# DF002 — Structural Validation

## Task Classification

- Investigation Task
- Criticidade: Medium
- Stakeholder principal: Head of Customer Retention

---

## Contexto Operacional

A diretoria executiva reportou aumento contínuo de churn nos últimos ciclos trimestrais.

Ainda não existe clareza sobre:

- perfis de maior risco
- comportamento dos clientes cancelados
- impacto de serviços adicionais
- possíveis padrões contratuais associados ao churn

Antes de qualquer modelagem, é necessário validar a integridade estrutural do dataset e produzir entendimento operacional inicial confiável.

---

## Objetivo da Task

Executar validação estrutural inicial completa do dataset.

O foco desta etapa NÃO é modelagem ou EDA aprofundada.

O objetivo principal é:

- entender a estrutura dos dados
- identificar inconsistências
- validar integridade básica
- investigar relações estruturais iniciais
- produzir leitura operacional preliminar

---

## Escopo Esperado

Nesta etapa NÃO devemos focar em:

- gráficos avançados
- hipóteses profundas
- modelagem
- feature engineering
- conclusões executivas definitivas

O foco deve permanecer em:

- entendimento estrutural
- qualidade dos dados
- coerência lógica
- validação operacional inicial

---

## Entregáveis Esperados

O notebook deve conter:

1. Shape do dataset
2. Tipos das colunas
3. Verificação de missing values
4. Verificação de duplicados
5. Cardinalidade básica das colunas categóricas
6. Identificação de possíveis inconsistências estruturais
7. Observações investigativas iniciais

---

## Pontos de Atenção

Investigar cuidadosamente:

- coluna `TotalCharges`
- categorias condicionais como:
  - `No internet service`
  - `No phone service`
- coerência lógica entre serviços contratados

---

## Diretrizes Operacionais

Não utilizar apenas outputs brutos.

Cada etapa deve possuir:

- interpretação curta
- observação operacional
- leitura investigativa inicial

Exemplo inadequado:

```python
df.isnull().sum()
```

Sem interpretação analítica.

Exemplo esperado:

```text
Não foram encontrados missing values explícitos, porém algumas colunas podem possuir ausência semântica mascarada por categorias como "No internet service".
```

---

## Critérios de Aprovação

A entrega será avaliada considerando:

- organização do notebook
- fluxo lógico
- clareza investigativa
- interpretações coerentes
- rigor estrutural
- rastreabilidade
- ausência de conclusões precipitadas

---

## Observação Final

Ainda não avançar para EDA aprofundada.

A prioridade atual é validar disciplina investigativa estrutural antes de etapas analíticas mais avançadas.

## 1. Contexto da Task

Neste momento iremos fazer uma análise limpa, ou seja, iremos focar na estrutura do dataset, verificar existência de valores nulos, tipos inconcistentes, existência de valores duplicados, etc. O foco aqui não é manipular dados, criar novas colunas nem nada, é apenas verificar o grau de confiabilidade que os dados nos oferecem.

## 2. Imports e Setups

In [1]:
import pandas as pd

print(f"Versão do pandas: {pd.__version__}")

Versão do pandas: 2.2.2


## 3. Load do Dataset

O dataset `Telco Customer Churn` já foi baixado na demanada anterior [Intake and data understanding](./01_intake_and_data_understanding.ipynb). Nosso objetivo agora é carregar ele para este ambiente.

In [2]:
try:
    df_origin = pd.read_csv(
        "../data/raw/telco-customer-churn/WA_Fn-UseC_-Telco-Customer-Churn.csv"
    )
except Exception as e:
    print(e)

df = df_origin.copy()

## 4. `head()` Inicial

In [3]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


O dataset mostra registros de clientes de uma operadora de telecomunicação. Os registros contam com dados referentes ao cliente e se o mesmo possui ou não alguns serviços disponíveis pela operadora. Uma coluna de valor pertinente é se o cliente cancelou ou não sua assinatura. O que pode nos ajudar a prever e/ou prevenir possíveis cancelamentos.

## 5. Shape do Dataset

In [4]:
registros, campos = df.shape

print(
    f"""
O Dataset em questão possui:
    - {registros} linhas
    - {campos} colunas
"""
)


O Dataset em questão possui:
    - 7043 linhas
    - 21 colunas



Ou melhor, nosso dataset possui 7043 registros de clientes. Cada cliente possui um total de 21 campos de informações referentes.

## 6. `info()` e Tipagem

É muito importante conhecermos os tipos de cada um dos campos de informação dos registros. Isso nos ajuda a nos familiarizar com os dados, encontrar inconcistências, perceber padrões ente outros. 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [6]:
colunas_numericas = df.select_dtypes(include="number").columns
colunas_categoricas = df.select_dtypes(include=["object", "category"]).columns

print(5 * "=" + "  Colunas Numéricas  " + 5 * "=" + "\n")
for col in colunas_numericas:
    print(f"{col}: {df[col].dtype}")

print("\n" + 5 * "=" + "  Colunas Categóricas  " + 5 * "=" + "\n")
for col in colunas_categoricas:
    print(f"{col}: {df[col].dtype}")

=====  Colunas Numéricas  =====

SeniorCitizen: int64
tenure: int64
MonthlyCharges: float64

=====  Colunas Categóricas  =====

customerID: object
gender: object
Partner: object
Dependents: object
PhoneService: object
MultipleLines: object
InternetService: object
OnlineSecurity: object
OnlineBackup: object
DeviceProtection: object
TechSupport: object
StreamingTV: object
StreamingMovies: object
Contract: object
PaperlessBilling: object
PaymentMethod: object
TotalCharges: object
Churn: object


Fazendo uma análise rápida me vem de primeira que o campo `SeniorCitizen` seguindo a lógica deveria ser do tipo object (Yes ou No). Isso partindo da ideia que as outras colunas segue este padrão. Porém acredito que para modelos o mais correto seria seguir padrão numérico de `SeniorCitizen`. Já as demais categóricas poderiam passar por uma transformação. Outra inconsistência que percebi é, por que `MonthlyCharges` é numérica e `TotalCharges` não?

A princípio o dataset não possui registros incompletos, com dados faltantes. Mas uma análise mais profunda pode ser necessária.

## 7. Investigação de `TotalCharges`

Assim como em todas as colunas com tipo object é importante uma análise mais profunda, pois as mesmas podem "esconder" valores nulos ou inconsistentes.

Uma primeira análise para verificar a existência de valores incosistentes pode ser tentar uma converção direta para `float64`.

In [7]:
try:
    df.TotalCharges.astype("float64")
except:
    print("Possível existência de valores inconsistentes!")

Possível existência de valores inconsistentes!


Irei aplicar uma validação mais direta. Transformar valores para float e aqueles que não puderem ser convertidos (que geram erros) transforemar em `NaN`.

In [8]:
tentativa_conversão = pd.to_numeric(df.TotalCharges, errors="coerce")
tentativa_conversão.isna().sum()

11

Parece que a coluna `TotalCharges` possui 11 valores inconsistentes. Aplicarei alguns métodos para encontrá-los.

In [9]:
df[tentativa_conversão.isna()]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,,No


Na verdade os valores estavam como strings vazias ou espaços.

## 8. Missing Values

Segundo `.info()` nosso dataset possui todos os valores, ou seja, não existe valores ausentes explicítos. Mas analisando a coluna `TotalCharges` verificamos 11 registros com essa coluna com valor "inconsistente", neste caso string vazia. Para as 3 colunas numéricas podemos confirmar que não possuem valores ausentes, já para as 18 colunas categóricas precisamos analisa uma a uma para verificar ou criar alguma função para automatizar processo.

## 9. Duplicados

Precisamos analisar o dataset em busca de regitros duplicados, ou seja, aqueles em que todos os valores são iguais. No caso de `customerID` iguais poderíamos considerar como duplicado mesmo os valores de seus campos sendo diferentes? Pois um cliente pode possuir mais de um contrato certo?

In [10]:
df.duplicated().sum()

0

Nosso dataset não possui registros idênticos. Vou verificar a existência de mais de um contrato com mesmo `customerID`

In [11]:
if df.customerID.nunique() == len(df):
    print("Não existe customerID duplicado")
else:
    print("Existe cliente com mais de um contrato")

Não existe customerID duplicado


## 10. Cardinalidade Categórica

O dataset possui 18 colunas categóricas, muitas das quais seria interessante transformação, mas neste momento estamos apenas analisando o grau de cardinalidade. Irei adotar 3 classificação, cardinalidade 2 chamarei de binária, entre 3 e 10 baixa , maior que 10 alta e maior que 30 irei averiguar melhor (exemplo `TotalCharges` deverá possuir alto valor de cardinalidade pois trata-se de valor final da conta).

In [12]:
colunas_categoricas

Index(['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService',
       'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'Contract', 'PaperlessBilling', 'PaymentMethod', 'TotalCharges',
       'Churn'],
      dtype='object')

In [13]:
def classificar_cardinalidade(valor):
    if valor == 2:
        return "Binária"
    elif 3 <= valor <= 10:
        return "Baixa"
    elif 11 <= valor <= 30:
        return "Moderada"
    else:
        return "Alta"

In [14]:
cardinalidade = pd.DataFrame(
    {"cardinalidade": [df[col].nunique() for col in colunas_categoricas]},
    index=colunas_categoricas,
)

cardinalidade["grau"] = cardinalidade.cardinalidade.map(
    classificar_cardinalidade
)
cardinalidade["valores"] = [
    (
        ", ".join(sorted(df[col].dropna().unique().tolist()))
        if cardinalidade.loc[col, "grau"] in ["Binária", "Baixa"]
        else None
    )
    for col in colunas_categoricas
]

In [15]:
cardinalidade

,cardinalidade,grau,valores
customerID,7043,Alta,None
gender,2,Binária,"Female, Male"
Partner,2,Binária,"No, Yes"
Dependents,2,Binária,"No, Yes"
PhoneService,2,Binária,"No, Yes"
MultipleLines,3,Baixa,"No, No phone service, Yes"
InternetService,3,Baixa,"DSL, Fiber optic, No"
OnlineSecurity,3,Baixa,"No, No internet service, Yes"
OnlineBackup,3,Baixa,"No, No internet service, Yes"
DeviceProtection,3,Baixa,"No, No internet service, Yes"


## 11. Relações Estruturais entre Colunas

Aqui precisamos verificar alguns pormenores, consigo observar 2. Primeiro as colunas onde `PhoneService` consta `no` a coluna `MultiplesLines` deve conter `No phone service`, já para coluna `InternetService` com valor `No` desencadeia que outras 6 colunas devem conter o valor `No internet service`.

In [16]:
# `PhoneService` -> No
sem_phone_service = df.loc[df.PhoneService == "No", ["MultipleLines"]]
sem_phone_service.nunique()

MultipleLines    1
dtype: int64

In [17]:
# `InternetService` -> No
sem_internet_service = df.loc[
    df.InternetService == "No",
    [
        "OnlineSecurity",
        "OnlineBackup",
        "DeviceProtection",
        "TechSupport",
        "StreamingTV",
        "StreamingMovies",
    ],
]
sem_internet_service.nunique()

OnlineSecurity      1
OnlineBackup        1
DeviceProtection    1
TechSupport         1
StreamingTV         1
StreamingMovies     1
dtype: int64

Todos os valores das colunas interligadas com `PhoneService` e `InternetService` no qual possuem valor 'No', constam com apenas um valor no qual está de acordo. Logo nosso dataset está com boa coerência lógica neste assunto.

## 12. Observações Investigativas Iniciais

O dataset parece estar com boa coerência lógica, com oucas inconsistências (encontrei apenas em `TotalCharges` onde os valores além de estarem com o tipo "object" possue 11 valores "nulos"). Pensei em fazer uma análise nas demais colunas categóricas mas ao calcular as cardinalidades de cada uma observei que não existia valores faltantes ou inconsistentes pois analisando as cardinalidades e os valoes possíveis fez todo sentido. Verificação que não fiz que talvez deveria ter feito é analisar a amplitude dos valores numéricos, assim como seus mínimos e máximos (não o fiz pois acredito que entraremos nisso mais adiante). Ao meu ver poderíamos mais adiante transformar algumas colunas categóricas, como por exemplo as que possuem cardinalidade binária em algo como Yes -> 1 e No -> 0, utilizar outras técnicas para as colunas com cardinalidade 3 e 4, para `TotalCharges` transformar em valores float.

No mais acredito ser um dataset com alto valor de estudo, bem estruturado, com coerência lógica inicial. Onde nosso trabalho possa ser focar como já dito em prever clientes que cancelariam o contrato, encontrar campos que de uma forma mais direta influenciam nesta decisão, etc.